src/inputs/mod.rs

```diff
@@ -78,6 +77,49 @@ impl InputConfig {
     }
 }
 
+// This trait has to be implemented by each InputConfig subclass.
+// It is used to create an InputReader from an InputConfig.
+pub(crate) trait InputBuilder {
```

Comment:
FWIW, I don't know that it's worth it to try to extract this builder pattern further to be used across config types. Like not all concrete Rust-side instances will need things like `resume_snapshot` to be built or `worker_count`, so it might end up feeling messy. I guess we can return to this later.

Reply:
The idea was to make the pattern generic over the arguments passed to the `build` function, with the `type Input;` type param.
This would be different for each module. In InputConfig it's `(WorkerIndex, usize, Option<StateBytes>)`, in OutputConfig it's `(WorkerIndex, usize)` and in ClockConfig it's `()`.

Each subclass of each module would have to share the params set, but that's already happening, so we should be able to have the same behavior.

Anyway I didn't include it in this PR because I too am not sure it's worth it, since it makes the code a bit weirder to read.